src/Language/TSX/Assignment.hs

```diff
@@ -0,0 +1,971 @@
+  , Syntax.Identifier
+  , Syntax.Context
+  , Type.Readonly
+  , Type.TypeParameters
+  , TSX.Syntax.TypeParameter
+  , TSX.Syntax.Constraint
+  , TSX.Syntax.ParenthesizedType
+  , TSX.Syntax.DefaultType
+  , TSX.Syntax.PredefinedType
+  , TSX.Syntax.TypeIdentifier
+  , TSX.Syntax.NestedIdentifier
+  , TSX.Syntax.NestedTypeIdentifier
+  , TSX.Syntax.GenericType
+  , TSX.Syntax.TypeArguments
+  , TSX.Syntax.TypePredicate
+  , TSX.Syntax.CallSignature
+  , TSX.Syntax.ConstructSignature
+  , TSX.Syntax.ArrayType
+  , TSX.Syntax.LookupType
+  , TSX.Syntax.FlowMaybeType
+  , TSX.Syntax.TypeQuery
+  , TSX.Syntax.IndexTypeQuery
+  , TSX.Syntax.ThisType
+  , TSX.Syntax.ExistentialType
+  , TSX.Syntax.AbstractMethodSignature
+  , TSX.Syntax.IndexSignature
+  , TSX.Syntax.ObjectType
+  , TSX.Syntax.LiteralType
+  , TSX.Syntax.Union
+  , TSX.Syntax.Intersection
+  , TSX.Syntax.Module
+  , TSX.Syntax.InternalModule
+  , TSX.Syntax.FunctionType
+  , TSX.Syntax.Tuple
+  , TSX.Syntax.Constructor
+  , TSX.Syntax.TypeAssertion
+  , TSX.Syntax.ImportAlias
+  , TSX.Syntax.Debugger
+  , TSX.Syntax.ShorthandPropertyIdentifier
+  , TSX.Syntax.Super
+  , TSX.Syntax.Undefined
+  , TSX.Syntax.ClassHeritage
+  , TSX.Syntax.AbstractClass
+  , TSX.Syntax.ImplementsClause
+  , TSX.Syntax.JsxElement
+  , TSX.Syntax.JsxSelfClosingElement
+  , TSX.Syntax.JsxOpeningElement
+  , TSX.Syntax.JsxText
+  , TSX.Syntax.JsxClosingElement
+  , TSX.Syntax.JsxExpression
+  , TSX.Syntax.JsxAttribute
+  , TSX.Syntax.JsxFragment
+  , TSX.Syntax.JsxNamespaceName
+  , TSX.Syntax.OptionalParameter
+  , TSX.Syntax.RequiredParameter
+  , TSX.Syntax.RestParameter
+  , TSX.Syntax.PropertySignature
+  , TSX.Syntax.AmbientDeclaration
+  , TSX.Syntax.EnumDeclaration
+  , TSX.Syntax.ExtendsClause
+  , TSX.Syntax.AmbientFunction
+  , TSX.Syntax.ImportRequireClause
+  , TSX.Syntax.ImportClause
+  , TSX.Syntax.LabeledStatement
+  , TSX.Syntax.Annotation
+  , TSX.Syntax.With
+  , TSX.Syntax.ForOf
+  , TSX.Syntax.Update
+  , TSX.Syntax.ComputedPropertyName
+  , TSX.Syntax.Decorator
+  , TSX.Syntax.Import
+  , TSX.Syntax.QualifiedAliasedImport
+  , TSX.Syntax.SideEffectImport
+  , TSX.Syntax.DefaultExport
+  , TSX.Syntax.QualifiedExport
+  , TSX.Syntax.QualifiedExportFrom
+  , TSX.Syntax.JavaScriptRequire
+  , []
+  , Statement.StatementBlock
+  , TSX.Syntax.MetaProperty
+  , TSX.Syntax.AnnotatedExpression
+  ]
+
+type Term = Term.Term (Sum Syntax) Location
+type Assignment = Assignment.Assignment [] Grammar
+
+-- | Assignment from AST in TSX’s grammar onto a program in TSX’s syntax.
+assignment :: Assignment Term
+assignment = handleError $ makeTerm <$> symbol Program <*> children (Statement.Statements <$> manyTerm statement) <|> parseError
+
+expression :: Assignment Term
+expression = handleError everything
+  where
+    everything = choice [
+      asExpression,
+      nonNullExpression',
+      importAlias',
+      internalModule,
+      super,
+      object,
+      array,
+      jsxElement',
```

Comment:
I've had to duplicate typescript assignment in order to leave `jsxElement'` and `jsxFragment` in here. I could move assignment out into some `Common` module and pass in a list of terms to mappend instead. Thoughts?

Reply:
Thanks for calling out the duplication. Given the size of this patch as it is, I'd be in favor of merging this and to extract a common assignment module as a follow up PR. That might also make a weightier issue for someone in the community to take on. Either way I'm 👍